README.md
Gabctk
======

Outil pour travailler sur les fichiers gabc.

[English Documentation](http://www.sspxusa.org/goodies/gabc2mid/). Note : this doc is for legacy gabc2mid ; it's obsolete on several points.

Gabc
----

Le *gabc* est le langage utilisé par le logiciel
[Gregorio](http://home.gna.org/gregorio/).
Vous en trouverez la description [ici](http://home.gna.org/gregorio/gabc/).


Gabctk
------

Ce script est dérivé de [gabc2mid](https://github.com/jperon/gabc2mid) ;
l'optique du projet s'étant élargie, gabc2mid restera tel qu'il est à présent
(sauf correction de bug), et les développements auront lieu ici.
Gabctk parcourt le code *gabc*, en extrait ce qui concerne la mélodie,
et produit celle-ci sous la forme d'un fichier midi et/ou lilypond, abc,
musicxml.
Il peut aussi extraire le texte dans un fichier texte.
La syntaxe est la suivante :

    gabctk.py -i </chemin/vers/le/fichier/source.gabc> \
             [-o </chemin/vers/le/fichier/destination.mid>] \
             [-l </chemin/vers/le/fichier/destination.ly>] \
             [-c </chemin/vers/le/fichier/destination.abc>] \
             [-x </chemin/vers/le/fichier/destination.xml>] \
             [-b </chemin/vers/le/fichier/destination.tab>] \
             [-e </chemin/vers/le/fichier/destination.txt>] \
             [-m </chemin/vers/le/fichier/destination.mus>] \
             [-t tempo] \
             [-d transposition] \
             [-a alerte] \

Toutes les options entre crochets sont facultatives.

Si, à la place d'un nom de fichier, vous voulez utiliser l'entrée ou la sortie
standard, spécifiez `-`. Par exemple, pour écouter un gabc grâce à `timidity` :

    gabctk.py -i <fichier/source.gabc> -o - | timidity -

Ou encore, pour extraire le texte du gabc et l'afficher :

    gabctk.py -i <fichier/source.gabc> -e -

Le tempo est exprimé en temps premiers par minute :
sa valeur par défaut est 165.

La transposition est exprimée en demi-tons. En son absence, gabctk transposera
automatiquement le chant sur une tessiture facile à chanter. Pour les formats
abc et musicxml, la gestion de la transposition est laissée à abc et aux
différents logiciels compatibles avec ces formats. Les notes resteront donc
graphiquement en place, mais la mélodie sera jouée à la hauteur indiquée par
ce paramètre.

Si des alertes sont définies, gabctk renverra un message chaque fois
qu'il détecte la chaîne de caractères dans le texte du chant.
Par exemple, `gabctk -i \<Fichier.gabc\> -a j -a eumdem` renverra un message
si le texte contient des *j* ou le mot *eumdem*.

Si vous ne voulez pas modifier les options par défaut,
vous pouvez aussi utiliser la syntaxe suivante :

    gabctk.py </chemin/vers/le/fichier/source.gabc>
             [</chemin/vers/le/fichier/destination.mid>]


abc2xml/Changes.txt
Version 71
- Added new option --meta to allow users to specify a mapping of ABC info fields to MusicXML meta data elements.
  Depending on the music editor used, some info fields may get a reasonable position onto the front page of the score
- The first T: infofield is translated to the MusicXML work-title element. Subsequent T: fields are concatenated onto
  the the MusicXML work-number element. In Finale and MuseScore the work-number is displayed in a smaller font
  below the work title.
- Multiple ABC info fields of the same type are concatenated with a new line as separator.
- When the -t option is used ontly the first T: info field is used as file name.

Version 70
- ABC meta data that is not supported in MusicXML is translated into <miscellaneous> elements.
  Previously this data was put into <credit> elements. However, <credit> elements are only meant for formatted display
  of data. Writing metadata without formatting into <credit> elements (as abc2xml did) inhibited display of other
  important data, like title, composer or lyricist.
- new option -t causes abc2xml to use the tune title as file name. Useful for multi tune files.

Version 69
- Remove redundant instrument definitions and tags from a multi voice part. Redundant instrument definitions can arise
  when ABC voices with identical midi instruments are mapped onto one (grand) staff.
- Duration of a tremolo pair is internally doubled, because MusicXML considers tremolo pairs as duplets (two notes in
  the time of one) and halves the duration again.
  Thus !//-!A2A2 is internally !//-!A4A4 and is translated as (2:1:2 A4A4 which has the same duration as A2A2 but the
  noteheads are those of an A4.

Version 68
- The vertical postition of a note specified with the staff step value in I:percmap, no longer depends on the current clef.
  the staff step now always refers to the vertical note position with a treble clef.
- When slurs close and open at the same note, the closing slurs are treated first.

Version 67
- %%MIDI instrument/channel changes translated to (text) directions in xml (works with Finale Notepad)
- implementation of I:percmap <abc-note> <staff-step> <midi-number> <xml-notehead> (Proposal by Alexander Scheutzow)
  See readme.txt for description of syntax, usage and semantics.
- %%MIDI drummap is translated to an equivalent I:percmap
- The octave= parameter was not recognized without a clef= in the same line (for both V: and K:)
- The octave= required a signed number +n of -n. Now the + sign is optional.
- translate stafflines attribute. (no correction for display-step yet)
- translation of %%MIDI rationalized.

Version 66
- deleted the code for the post-fix meaning of closing decorations. A decoration now always refers to the following
  note (even if it is in the next measure). The only exception is the slur closing bracket.
- replace illegal values of L: with default values (caused exception)
- replace illegal values of Q: with default values (caused exception)
- correctly parse multiple fields of V: (some fields were discarded)

Version 65
- multiple note tremolo decoration translated: !/-!, !//-! and !///-! (see readme.txt for details)
- an xml metronome element is only generated when a numeric tempo is present (e.g. Q: 1/4=108).
- when a Q: field only contains text, a hidden tempo value is derived (using tempo map from abc2midi).
- removed special grammar for decorations of chord notes
- removed special grammar for decorations before barline. The position of segno, coda and fine is now always at the
  correct side of the measure in which they are defined, independent of their ABC-position within that measure.
- move closing decorations in front of the note they belong to (postfix to prefix).

Version 64
- single tremolo decoration translated: !/!, !//! and !///!
- glissando decoration translated: !-(! ... !-)! and !~(! ... !~)! (Proposal by Alexander Scheutzow)
- added translations for !plus! or !+!
- text in Q: fields is added as text to the xml tempo direction.

Version 63
- changed the pageformat defaults to those of abcm2ps (A4) in stead of those used in MuseScore.
- in the -p command line option (page format) changed the "space" parameter (MuseScore) to the "scale" parameter (abcm2ps)
- python version test also works with python2.6 (caused an exeption previously).

Version 62
- mxl output was broken (option -z)
- clefs could be set to the wrong type when a type designation happened to occur in a name="..." string.
- small code change for compatibility with EasyABC's interface to abc2xml.

Version 61
- program can run with either Python 2.6, 2.7 or Python 3.x. (Python 3.4 runs 17% slower than 2.7 on my regression set)

Version 60
- use lyric alignment as defined in the 2.1 standard.
  Lyric lines are parsed as normal inlined information fields. (suggested by Alexander Scheutzow).
- integrate lyric grammar into the main abc grammar (was a separate grammer previously).
- only base name of input path is used when combined with the -o directory path. (correction P.Villiger)
- skip illegal user defined symbols (with warning) in stead of raising a syntax exception.

Version 59
- allow X: without number or with arbitrary textstring
- allow Y: information field
- the -r option is consequently passed as parameter to keep functions referentially transparent. (was a global variable)
- corrected error message for illegal use of U:

Version 58
- grammar for I:score instruction replaces ad hoc parsing.
- voice merging, part grouping and grand staffs are derived by transformation of the parse tree of I:score.
- also move clef redefinitions after [I:staff +/-n].
- avoid a false beam break when [I:staff +/-n] is used within a beamed note group.
- absolute staff numbers in [I:staff n] are mapped to MusicXML grand staff numbers (when possible) for reasons

windows/compilationsource/versioninfo.py
VSVersionInfo(
  ffi=FixedFileInfo(
    filevers=(0, 0, 1, 3),
    prodvers=(0, 0, 1, 3),
    mask=0x3f,
    flags=0x0,
    OS=0x40004,
    fileType=0x1,
    subtype=0x0,
    date=(0, 0)
    ),
  kids=[
    StringFileInfo(
      [
      StringTable(
        u'040904B0',
        [StringStruct(u'CompanyName', u'SSPX by Father Jaques Peron'),
        StringStruct(u'FileDescription', u'gabc2mid'),
        StringStruct(u'FileVersion', u'0.0.1.3'),
        StringStruct(u'InternalName', u'gabc2mid'),
        StringStruct(u'LegalCopyright', u'\xa9 SSPX All rights reserved.'),
        StringStruct(u'OriginalFilename', u'gabc2mid.exe'),
        StringStruct(u'ProductName', u'gabc2mid'),
        StringStruct(u'ProductVersion', u'0.0.1.3')])
      ]), 
    VarFileInfo([VarStruct(u'Translation', [1033, 1200])])
  ]
)

gabc2mid.py
#! /usr/bin/python3
# -*- coding: UTF-8 -*-

import os,sys,getopt
import re
from midiutil.MidiFile3 import MIDIFile

def gabc2mid(arguments):
    tempo = 165
    entree = sortie = ''
    debug = False
    transposition = ''
    alertes = corrections = []
    try:
        opts, args = getopt.getopt(arguments,"hi:o:e:t:d:a:v",["help","entree=","sortie=","export=","tempo=","transposition=","alerter=","verbose"])
    except getopt.GetoptError:
        aide(1)
    for opt, arg in opts:
        if opt == '-h':
            aide(0)
        elif opt in ("-i", "--entree"):
            entree = FichierTexte(arg)
            sortie = Fichier(re.sub('.gabc','.mid',arg))
        elif opt in ("-o", "--sortie"):
            sortie = Fichier(arg)
        elif opt in ("-e", "--export"):
            texte = FichierTexte(arg)
        elif opt in ("-t", "--tempo"):
            tempo = int(arg)
        elif opt in ("-d", "--transposition"):
            transposition = int(arg)
        elif opt in ("-a", "--alerter"):
            alertes.append(arg)
        elif opt in ("-v", "--verbose"):
            debug = True
    try:
        if entree == '':
            entree = FichierTexte(arguments[0])
        if sortie == '':
            sortie = Fichier(re.sub('.gabc','.mid',arguments[0]))
    except IndexError: aide(2)
    try: gabc = Gabc(entree.contenu)
    except FileNotFoundError: aide(3)
    partition = Partition(gabc = gabc.musique, transposition = transposition)
    print(Note(hauteur = partition.tessiture['minimum']).nom
        + " - "
        + Note(hauteur = partition.tessiture['maximum']).nom)
    if debug:
        print(gabc.partition)
        print(partition.texte)
        print(partition.notes)
    midi = Midi(partition.notes,tempo)
    midi.ecrire(sortie.chemin)
    try: partition.verifier(alertes)
    except: pass
    try: texte.ecrire(partition.texte)
    except UnboundLocalError: pass

def aide(code):
    print('gabc2mid.py -i <input.gabc> [-o <output.mid>] [-e <texte.txt>] [-t <tempo>] [-v]')
    sys.exit(code)

class Gabc:
    def __init__(self,contenu):
        self.contenu = contenu
    @property
    def partition(self):
        resultat = self.contenu
        regex = re.compile('%%\n')
        resultat = regex.split(resultat)[1]
        resultat = re.sub('%.*\n','',resultat)
        resultat = re.sub('\n',' ',resultat)
        return resultat
    @property
    def musique(self):
        resultat = []
        partition = self.partition
        regex = re.compile('[cf][b]?[1234]')
        cles = regex.findall(partition)
        partiestoutes = regex.split(partition)
        parties = partiestoutes[0] + partiestoutes[1], partiestoutes[2:]
        for i in range(len(cles)):
            cle = cles[i]
            try:
                for n in parties[i]:
                    resultat.append((cle,n))
            except IndexError:
                sys.stderr.write("Il semble que vous ayez des changements de clé sans notes subséquentes. Le résultat n'est pas garanti.\n")
        return resultat

class Partition:
    def __init__(self,**parametres):
        self.b = ''
        transposition = None
        if 'partition' in parametres:
            self.notes = parametres['pitches']
        if 'bemol' in parametres:
            self.b = self.b + parametres['bemol']
        if 'gabc' in parametres:
            self.notes,self.texte = self.g2p(parametres['gabc'])
        if 'transposition' in parametres:
            try: transposition = int(parametres['transposition'])
            except ValueError: pass
        self.transposer(transposition)
    def g2p(self,gabc):
        gamme = "abcdefghijklm"
        episeme = '_'
        point = '.'
        quilisma = 'w'
        speciaux = 'osvOSV'
        barres = '`,;:'
        bemol = "x"
        becarre = "y"
        coupures = '/ '
        notes = []
        b = '' + self.b
        mot = 0
        texte = ''
        neume = 0
        neumeencours = ''
        musique = 0
        for i in range(len(gabc)):
            signe = gabc[i]
            if musique == 1:
                if signe[1].lower() in gamme:
                    j = 0
                    s = 0
                    note = Note(gabc = signe, bemol = b)
                    notes.append(note)
                    memoire = signe
                elif signe[1] in speciaux:
                    s += 1
                    if s > 1:
                        note = Note(gabc = memoire, bemol = b)
                        notes.append(note)
                elif signe[1] == episeme:
                    neumeencours = neume
                    j -= 1
                    notes[j].duree = 1.7
                elif signe[1] == point:
                    j -= 1
                    notes[j].duree = 2.3
                elif signe[1] == quilisma:
                    notes[-2].duree = 2
                elif signe[1] == bemol:
                    b = b + memoire[1]
                    notes = notes[:-1]
                elif signe[1] == becarre:
                    re.sub(memoire[1],'',b)
                    notes = notes[:-1]
                elif signe[1] in coupures:
                    if notes[-1].duree < notes[-2].duree:
                        notes[-1].duree = notes[-2].duree
                elif signe[1] in barres or signe[1] in coupures:
                    b = '' + self.b
                    if signe[1] == ';':
                        notes[-1].duree += .5
                    elif signe[1] == ':':
                        notes[-1].duree += 1
                else:
                    if signe[1] == ')':
                        musique = 0
                    if signe[1] == '[':
                        musique = 2
                    if neumeencours == neume and notes[-1].duree < notes[-2].duree:
                        notes[-1].duree = notes[-2].duree
            elif musique == 0:
                if signe[1] == '(':
                    musique = 1
                    neume += 1
                elif signe[1] in ('{', '}'): pass
                else:
                    if signe[1] == ' ':
                        mot += 1
                        b = '' + self.b
                        try:
                            if texte[-1] != ' ':
                                texte += signe[1]
                        except IndexError: pass
                    else: texte += signe[1]
            elif musique == 2:
                if signe[1] == ']':
                    musique = 1
        return notes, texte
    def transposer(self,transposition):
        if transposition == None: t = 66 - int(sum(self.tessiture.values())/2)
        else: t = transposition
        for i in range(len(self.notes)):
            self.notes[i].hauteur += t
    @property
    def tessiture(self):
        minimum = maximum = 0
        for note in self.notes:
            if minimum == 0 or note.hauteur < minimum: minimum = note.hauteur
            if note.hauteur > maximum: maximum = note.hauteur
        return {'minimum': minimum, 'maximum': maximum}
    def verifier(self,alertes):
        for alerte in alertes:
            if alerte in self.texte:
                print("!!! " + alerte + " !!!")

class Note:
    def __init__(self,**parametres):
        self.b = ''
        if 'bemol' in parametres:
            self.b = parametres['bemol']
        if 'hauteur' in parametres:
            self.hauteur = parametres['hauteur']
        if 'duree' in parametres:
            self.duree = parametres['duree']
        if 'gabc' in parametres:
            self.hauteur,self.duree = self.g2p(parametres['gabc'])
    @property
    def nom(self):
        o = int(self.hauteur / 12) - 1
        n = int(self.hauteur % 12)
        return ('Do','Do#','Ré','Ré#','Mi','Fa','Fa#','Sol','Sol#','La','La#','Si')[n] + str(o)
    def g2p(self,gabc):
        la = 57
        si = la + 2
        do = la + 3
        re = do + 2
        mi = re + 2
        fa = mi + 1
        sol = fa + 2
        octave = 12
        cle = gabc[0]
        if len(cle) == 3:
            cle = cle[0] + cle[2]
            si = la + 1
        note = gabc[1]
        decalage = {
            "c4": 0,
            "c3": 2,
            "c2": 4,
            "c1": 6,
            "f4": 3,
            "f3": 5,
            "f2": 0,
            "f1": 2
        }
        gamme = (la, si, do, re, mi, fa, sol)
        i = decalage[cle] - 1
        o = 0
        if cle == 'f3': o = -12
        notes = {}
        for j in "abcdefghijklm":
            try:
                i += 1
                notes[j] = gamme[i] + o
            except IndexError:
                i -= 7
                o += 12
                notes[j] = gamme[i] + o
        duree = 1
        h = note.lower()
        hauteur = notes[h]
        if h in self.b:
            hauteur -= 1
        return hauteur,duree

class Midi:
    def __init__(self,partition,tempo):
        piste = 0
        temps = 0
        self.sortieMidi = MIDIFile(1)
        self.sortieMidi.addTrackName(piste,temps,"Gregorien")
        self.sortieMidi.addTempo(piste,temps, tempo)
        self.sortieMidi.addProgramChange(piste,0,temps,74)
        for note in partition:
            channel = 0
            pitch = note.hauteur
            duree = note.duree
            volume = 127
            self.sortieMidi.addNote(piste,channel,pitch,temps,duree,volume)
            temps += duree
    def ecrire(self,chemin):
        binfile = open(chemin, 'wb')
        self.sortieMidi.writeFile(binfile)
        binfile.close()

class Fichier:
    def __init__(self,chemin):
        self.dossier = os.path.dirname(chemin)
        self.nom = os.path.splitext(os.path.basename(chemin))[0]
        self.chemin = chemin

class FichierTexte:
    def __init__(self,chemin):
        self.dossier = os.path.dirname(chemin)
        self.nom = os.path.splitext(os.path.basename(chemin))[0]
        self.chemin = chemin
    @property
    def contenu(self):
        fichier = open(self.chemin,'r')
        texte = fichier.read(-1)
        fichier.close()
        return texte
    def ecrire(self,contenu):
        fichier = open(self.chemin,'w')
        fichier.write(contenu)
        fichier.close()

if __name__ == '__main__':
    gabc2mid(sys.argv[1:])
